Sources/SwiftProtobuf/SwiftProtobufError.swift

```diff
@@ -0,0 +1,620 @@
+        }
+        
+        public static var binaryDecodingError: Self {
+            Self(.binaryDecodingError)
+        }
+        
+        public static var jsonEncodingError: Self {
+            Self(.jsonEncodingError)
+        }
+        
+        public static var jsonDecodingError: Self {
+            Self(.jsonDecodingError)
+        }
+        
+        public static var textFormatDecodingError: Self {
+            Self(.textFormatDecodingError)
+        }
+        
+        public static var anyUnpackError: Self {
+            Self(.anyUnpackError)
+        }
+        
+        public static var protoFileToModuleMappingError: Self {
+            Self(.protoFileToModuleMappingError)
+        }
+    }
+
+    /// A location within source code.
+    public struct SourceLocation: Sendable, Hashable {
+        /// The function in which the error was thrown.
+        public var function: String
+
+        /// The file in which the error was thrown.
+        public var file: String
+
+        /// The line on which the error was thrown.
+        public var line: Int
+
+        public init(function: String, file: String, line: Int) {
+            self.function = function
+            self.file = file
+            self.line = line
+        }
+
+        @usableFromInline
+        internal static func here(
+            function: String = #function,
+            file: String = #fileID,
+            line: Int = #line
+        ) -> Self {
+            return SourceLocation(function: function, file: file, line: line)
+        }
+    }
+}
+
+extension SwiftProtobufError: CustomStringConvertible {
+    public var description: String {
+        "\(self.code): \(self.message)"
+    }
+}
+
+extension SwiftProtobufError: CustomDebugStringConvertible {
+    public var debugDescription: String {
+        "\(String(reflecting: self.code)): \(String(reflecting: self.message))"
+    }
+}
+
+// - MARK: Common errors
+
+extension SwiftProtobufError {
+    /// Errors arising from encoding protobufs into binary data.
+    public enum BinaryEncoding {
+        /// The definition of the message or one of its nested messages has required
+        /// fields but the message being encoded did not include values for them. You
+        /// must pass `partial: true` during encoding if you wish to explicitly ignore
+        /// missing required fields.
+        public static let missingRequiredFields = SwiftProtobufError(
+            code: .binaryEncodingError,
+            message: """
+                The definition of the message or one of its nested messages has required fields, \
+                but the message being encoded did not include values for them. \
+                Decode with `partial: true` if you wish to explicitly ignore missing required fields.
+            """
+        )
+        
+        /// Messages are limited to a maximum of 2GB in encoded size.
+        public static let tooLarge = SwiftProtobufError(
+            code: .binaryEncodingError,
```

Comment:
We really shouldn't use a single `code` for more than one error otherwise users can't match on the code. This means introducing a ton more cases in the underlying `code` enum but giving developers back better error matching.

Also all these let's should probably be `func`s if we want to pass through the `file/line/column` since these need to be created at the place where the error is created and not here where it is statically defined

Reply:
Ah, I was using [`NIOFileSystem/FileSystemError`](https://github.com/apple/swift-nio/blob/fc63f0cf4e55a4597407a9fc95b16a2bc44b4982/Sources/NIOFileSystem/FileSystemError.swift#L29) as inspiration, where we only have a few `Code`s that group the error types in different categories, and then throw errors with different messages but the appropriate code.

Not sure I follow why users wouldn't be able to match on the code. This is possible:
```swift
do {
  try someOperationThatThrows()
} catch let error as SwiftProtobufError where error.code == .binaryEncodingError {
  // Do something
}
```
And so is this:
```swift
do {
  try someOperationThatThrows()
} catch let error as SwiftProtobufError {
  switch error.code {
  case .binaryEncodingError:
    // Do something
  case ...
  }
}
```

Regarding these having to be `func`s though, you're right, I was unsure about keeping the `SourceLocation` and did a refactor and I'm not really using it anyways but didn't properly clean up, my bad. I see there's another thread above discussing whether we want to keep it or not. If we do, I'll convert all of these into `func`s and add the location to the message. Otherwise I'll just get rid of it and keep these as `let`s.